src/util/util_watchdog.c

```diff
@@ -41,20 +41,29 @@ static bool watchdog_detect_timeshift(void)
 {
     time_t prev_time;
     time_t cur_time;
-    errno_t ret;
 
     prev_time = watchdog_ctx.timestamp;
     cur_time = watchdog_ctx.timestamp = time(NULL);
     if (cur_time < prev_time) {
-        /* Time shift detected. We need to restart watchdog. */
-        DEBUG(SSSDBG_IMPORTANT_INFO, "Time shift detected, "
-              "restarting watchdog!\n");
-        teardown_watchdog();
-        ret = setup_watchdog(watchdog_ctx.ev, watchdog_ctx.input_interval);
-        if (ret != EOK) {
-            DEBUG(SSSDBG_FATAL_FAILURE, "Unable to restart watchdog "
-                  "[%d]: %s\n", ret, sss_strerror(ret));
-            orderly_shutdown(1);
+        struct itimerspec its;
+        errno_t ret;
+        /* Time shift detected.
+         *
+         * As restarting the watchdog from the signal_handler may
+         * lead sssd_be to a deadlock due to the use of non async-
+         * signal-safe function. let's just (re)set the timer's
+         * time.
+         */
+
+        /* As done when setting up the watchdog handler,
+         * we give 1 second head start to the watchdog event */
+        its.it_value.tv_sec = watchdog_ctx.interval.tv_sec + 1;
+        its.it_value.tv_nsec = 0;
+        its.it_interval.tv_sec = watchdog_ctx.interval.tv_sec;
+        its.it_interval.tv_nsec = 0;
+        ret = timer_settime(watchdog_ctx.timerid, 0, &its, NULL);
+        if (ret == -1) {
+            _exit(1);
```

Comment:
This is not sufficient because we need to reset the tevent timer. I think watchdog itself is actually fired in the right time because the timer clock is adjusted. The problematic timer is the tevent one, which needs to be rescheduled.